src/placeholder.ts

```diff
@@ -0,0 +1,536 @@
+          fenceMarkerCode = code;
+          fenceMarkerLength = j - i;
+          i = k - 1;
+          continue;
+        }
+      }
+    }
+
+    // Skip fenced code blocks (``` or ~~~).
+    if (fenceMarkerCode !== 0) continue;
+
+    // Skip code spans. Code spans are terminated either by the same number of
+    // backticks (``foo``) or by a blank line (only spaces).
+    if (state === STATE_DATA && code === CODE_BACKTICK) {
+      const j = skipCode(input, code, i + 1);
+      for (let k = j; k < n; ++k) {
+        switch (input.charCodeAt(k)) {
+          case code: {
+            const l = skipCode(input, code, k);
+            if (l - k === j - i) {
+              // paired terminator
+              i = l - 1;
+              continue out;
+            } else {
+              // skip unpaired terminator
+              k = l - 1;
+            }
+            break;
+          }
+          case CODE_LF: {
+            const l = skipSpace(input, k + 1);
+            if (input.charCodeAt(l) === CODE_LF) {
+              // blank line
+              i = l - 1;
+              continue out;
+            }
+            break;
+          }
+        }
+      }
+    }
+
+    // Detect inline expressions.
+    if (state === STATE_DATA) {
+      if (code === CODE_BACKSLASH) {
+        afterBackslash = true;
+      } else if (code === CODE_DOLLAR) {
+        if (afterBackslash) {
+          content += input.slice(index, (index = i) - 1);
+          afterBackslash = false;
+        } else {
+          afterDollar = true;
+        }
+      } else if (afterBackslash) {
+        afterBackslash = false;
+        if (afterDollar && code === CODE_LBRACE) {
+          content += input.slice(index, (index = i) - 1);
+        }
+      } else if (afterDollar) {
+        afterDollar = false;
+        if (code === CODE_LBRACE) {
+          const parser = new (Parser as any)(acornOptions, input, i + 1); // private constructor
+          let braces = 1;
+          try {
+            do {
+              parser.nextToken();
+              if (parser.type === tokTypes.braceL || parser.type === tokTypes.dollarBraceL) {
+                ++braces;
+              } else if (parser.type === tokTypes.braceR && !--braces) {
+                if ((content += input.slice(index, i - 1))) yield {type: "content", value: content}, (content = "");
+                yield {type: "code", value: input.slice(i + 1, (i = parser.pos - 1))};
+                index = parser.pos;
+                break;
+              }
+            } while (parser.type !== tokTypes.eof);
+          } catch (error) {
+            if (!(error instanceof SyntaxError)) throw error;
+            // on invalid token (e.g., unterminated template, invalid unicode escape),
+            // read until the braces are closed
+            let j = parser.pos;
+            for (; j < n; ++j) {
+              if (input.charCodeAt(j) === CODE_RBRACE && !--braces) {
```

Comment:
Maybe we should just stop at the first close brace here? Because we aren’t incrementing in left brace anymore so it feels weird to still be counting. And it’s a degenerate case anyway so probably we want to do the simplest thing.

Reply:
Simplified by now just ignoring inline expressions if the tokens are invalid.